### PR TITLE
4476 rnr dirty state

### LINKED ISSUE #4476: Manage R&R form dirty state
## Is your feature request related to a problem? Please describe 👀

When I create an R&R form, if I'm happy with all the pre-populated values, I want to simply finalise the form. (Currently I have to click each line to confirm it before I can finalise)

When I edit some data and then navigate away, I want to be prompted to save my changes

When I finalise the form, if I have made changes to some lines but have not confirmed them, I want to be prompted before finalising.

## Describe the solution you'd like 🎁

R&R form needs to keep track of dirty state of the lines (possibly via a matrix? saving one line shouldn't reset dirty state to false, because other lines could be dirty?)

### PR DESCRIPTION
**Pull request by lache-melvin**
Fixes #4476

# 👩🏻‍💻 What does this PR do?

Adds dirty state to the R&R form. Changes include:

Indicate dirty lines to user:
Screenshot 2024-08-08 at 1 56 25 PM

Confirm on leaving if any lines are dirty:
Screenshot 2024-08-08 at 1 56 15 PM

Users must confirm dirty lines before finalising:
Screenshot 2024-08-08 at 1 56 37 PM

If lines are unconfirmed, but with no changes, confirm this with modal, then finalise:
Screenshot 2024-08-08 at 1 56 08 PM

## 💌 Any notes for the reviewer?

# 🧪 Testing

- [ ] Create R&R
- [ ] You can navigate away from the detail view
- [ ] Make some changes to at least one line
- [ ] Blue dot shows in the confirm column, on the line you changed
- [ ] You are now prompted when you try to navigate away
- [ ] Changes across multiple lines - saving one line, the other lines still show as unsaved changes
- [ ] Try to finalise, you get a message that you need to confirm changed lines
- [ ] Confirm those lines, but leave some untouched lines unconfirmed
- [ ] Click finalise - there is a confirmation modal warning that not all lines are confirmed
- [ ] Click OK - all lines are now checked as confirmed, and the form is finalised
- [ ] Create another R&R form, mark all lines as confirmed and finalise
- [ ] You get a confirmation modal, but no message about unconfirmed lines

# 📃 Documentation

- [x] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

File: client/packages/common/src/utils/environment/EnvUtils.ts [hunk 1]
```diff
@@ -96,7 +96,7 @@ const mapRoute = (route: string): RouteMapping => {
     case inRoute(AppRoute.ImmunisationPrograms):
       return { title: 'immunisations', docs: '/programs/immunizations/' };
     case inRoute(AppRoute.RnRForms):
-      return { title: 'r-and-r-forms', docs: '/programs/r-and-r-forms/' };
+      return { title: 'r-and-r-forms', docs: '/replenishment/r-and-r-forms/' };
```
Comment (lache-melvin) [inline client/packages/common/src/utils/environment/EnvUtils.ts:99]: Side quest: fix after moving the R&R forms to replenishment section

File: client/packages/requisitions/src/RnRForms/DetailView/Footer.tsx [hunk 2]
```diff
@@ -50,10 +67,7 @@ export const Footer = ({
             height={64}
           >
             <Box flex={1} display="flex" justifyContent="flex-end" gap={2}>
-              <DialogButton
-                onClick={() => navigateUpOne()}
-                variant={'cancel'}
-              />
+              <DialogButton onClick={() => navigateUpOne()} variant={'close'} />
```
Comment (lache-melvin) [inline client/packages/requisitions/src/RnRForms/DetailView/Footer.tsx:70]: Seeing as this just navigates back to the list view, have changed the button to say close... Not sure whether we actually want it to be a cancel/discard changes button?

Comment (andreievg) [reply client/packages/requisitions/src/RnRForms/DetailView/Footer.tsx:70]: Close sounds good to me. Could change it based on dirty state `close and discard`, but I think close works for now. Would be great to have this as a UI pattern, recorded somewhere

**Review by andreievg — COMMENTED** [review]
Changes look great, thank you.

Hmm i am getting `RnRFormAlreadyFinalised`

Change some lines, confirm then (not all lines), and then press the confirm button

File: client/packages/requisitions/src/RnRForms/api/hooks/useRnRForm.ts [hunk 3]
```diff
@@ -40,10 +40,22 @@ export const useRnRForm = ({ rnrFormId }: { rnrFormId: string }) => {
...
 
+  const updateLine = async (line: RnRFormLineFragment) => updateLines([line]);
+
+  const confirmRemainingLines = async () => {
+    if (!query.data) return;
+
+    let lines = query.data.lines
+      .filter(line => !line.confirmed)
+      .map(line => ({ ...line, confirmed: true }));
+    updateLines(lines);
```
Comment (andreievg) [inline client/packages/requisitions/src/RnRForms/api/hooks/useRnRForm.ts:51]: Await here i think

**Comment by lache-melvin** [comment]
> Changes look great, thank you.
> 
> Hmm i am getting `RnRFormAlreadyFinalised`
> 
> Change some lines, confirm then (not all lines), and then press the confirm button

I can't replicate this? Could you try again?

**Comment by andreievg** [comment]
> I can't replicate this? Could you try again? Or let me know what I'm missing?

I tried on the commit before the merge and i get this error,  after merge of 2.2.0 works great, thanks !

**Review by andreievg — APPROVED** [review]

**Pull request merged**